### PR TITLE
Replace time.Now() with clock.Now() in pkg/workload

### LINKED ISSUE #3380: Replace `time.Now()` code with `clock.Clock.Now()`
**What would you like to be added**:
There are packages like e.g. `pkg/scheduler` that use `time.Now()`. As a result this code is hard to test and prone to flakiness as we are not able to directly control the time during tests. It should be replaced with `clock.Clock.Now()` as it is  in WorkloadController

**Why is this needed**:

**Completion requirements**:

This enhancement requires the following artifacts:

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by TusharMohapatra07**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Replaces `time.Now()` to use the Now() function present inside `clock` package.

#### Which issue(s) this PR fixes:
Part of #3380

#### Special notes for your reviewer:
This PR is work-in-progress. I've refactored the code inside `pkg/workload` for now. I shall proceed to refactor more post review.

```release-note
NONE
```

### CODE REVIEW

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Please review.`preemptor` and `workloadReconciler` already had clock.Clock in them.

**Comment by mbobrovskyi** [comment]
What about this https://github.com/kubernetes-sigs/kueue/pull/3498/files#diff-c703a3b5f20a31c414c5cafc7df771810f8e39ba1e8658d8b9f897bcb8af659cL110?

**Comment by mbobrovskyi** [comment]
/retitle Replace time.Now() with clock.Now() in pkg/workload

**Comment by mbobrovskyi** [comment]
@TusharMohapatra07 Please use `/kind cleanup` in description.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Is the implementation correct ?  I've just refactored one function. After your approval, i'll do the rest.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi Is the implementation correct ? I've just refactored one function. After your approval, i'll do the rest.

LGTM.

**Comment by mbobrovskyi** [comment]
Also check tests. Maybe missed something.

**Comment by mbobrovskyi** [comment]
I think no necessary to add cleanup on the title. It's understandable by label.

**Comment by mbobrovskyi** [comment]
Also rename `Fixes #3380` to `Part of #3380`

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Yeah, that was in the readme template that we get when raising a PR

**Comment by mbobrovskyi** [comment]
Not sure, I think add clock to workloadInfo is better. WDYT?

**Comment by TusharMohapatra07** [comment]
I mean, we could add it but won't it break anything ? https://github.com/kubernetes-sigs/kueue/blob/0f54466b991a7d198eeca306dc75f2aee518c482/pkg/controller/admissionchecks/provisioning/controller.go#L98 we don't have such function for `workloadInfo` where we could pass in `realClock`

**Comment by mbobrovskyi** [comment]
I think we can set it here? https://github.com/kubernetes-sigs/kueue/blob/0f54466b991a7d198eeca306dc75f2aee518c482/pkg/controller/admissionchecks/provisioning/controller.go#L147-L149

**Comment by mbobrovskyi** [comment]
On the other hand we should propagate also to https://github.com/kubernetes-sigs/kueue/blob/0f54466b991a7d198eeca306dc75f2aee518c482/pkg/controller/admissionchecks/provisioning/controller.go#L492

**Comment by TusharMohapatra07** [comment]
> On the other hand we should propagate also to
> 
> https://github.com/kubernetes-sigs/kueue/blob/0f54466b991a7d198eeca306dc75f2aee518c482/pkg/controller/admissionchecks/provisioning/controller.go#L492

In the sense, pass into the function ? or just pass clock.Now() inside `SetAdmissionCheckState` ?

**Comment by mbobrovskyi** [comment]
> > On the other hand we should propagate also to
> > https://github.com/kubernetes-sigs/kueue/blob/0f54466b991a7d198eeca306dc75f2aee518c482/pkg/controller/admissionchecks/provisioning/controller.go#L492
> 
> In the sense, pass into the function ? or just pass clock.Now() inside `SetAdmissionCheckState` ?

I think `now time.Time` is enough.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi then i think we can eliminate the need of defining clock inside the `workloadInfo` struct. WDTY ??

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi then i think we can eliminate the need of defining clock inside the `workloadInfo` struct. WDTY ??

Both solutions work for me.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi or we could do this 
```
func (wlInfo *workloadInfo) update(wl *kueue.Workload, now time.Time) {
	for _, check := range wl.Status.AdmissionChecks {
		workload.SetAdmissionCheckState(&wlInfo.checkStates, check, now)
	}
	wlInfo.requeueState = wl.Status.RequeueState
}

func (c *Controller) syncCheckStates(
	ctx context.Context, wl *kueue.Workload,
	wlInfo *workloadInfo,
	checkConfig map[string]*kueue.ProvisioningRequestConfig,
	activeOrLastPRForChecks map[string]*autoscaling.ProvisioningRequest,
) error {
	log := ctrl.LoggerFrom(ctx)
	wlInfo.update(wl, c.clock.Now()) //we can use the clock on controller
        ....
```

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 1]
```diff
@@ -442,6 +445,7 @@ func newWlReconciler(c client.Client, helper *multiKueueStoreHelper, cRec *clust
 		deletedWlCache:    utilmaps.NewSyncMap[string, *kueue.Workload](0),
 		eventsBatchPeriod: eventsBatchPeriod,
 		adapters:          adapters,
+		clock:             realClock,
```
Comment (mbobrovskyi) [inline pkg/controller/admissionchecks/multikueue/workload.go:448]: We can't use `FakeClock` this way, so this solution has no benefit—`realClock.Now()`and `time.Now()` are essentially the same.

Since we aren’t using Option, maybe we should just pass it directly to the constructor?

Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: You mean something like func newWlReconciler(c client.Client, helper *multiKueueStoreHelper, cRec *clustersReconciler, origin string, workerLostTimeout, eventsBatchPeriod time.Duration, adapters map[string]jobframework.MultiKueueAdapter, clk clock.Clock) *wlReconciler ??

Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: Yeah, something like that I think.

Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: Does it work?

Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: On the other hand we also can implement Options here.

Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: I think with Options solution is better but takes more changes.

Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/multikueue/workload.go:448]: If it's better then i have no issue in implementing that. I think `pkg/scheduler` has it. I just have to do the same thing here.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi I've added options and used fakeClock in the testing, Please check.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi I've added options and used fakeClock in the testing, Please check.

Please fix tests.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi How do i exactly do that ? How will i know which tests are failing ?

File: pkg/controller/admissionchecks/multikueue/workload_test.go [hunk 2]
```diff
@@ -50,6 +51,9 @@ var (
 )
 
 func TestWlReconcile(t *testing.T) {
+	now := time.Now()
+	fakeClock := testingclock.NewFakeClock(now)
```
Comment (mbobrovskyi) [inline pkg/controller/admissionchecks/multikueue/workload_test.go:55]: If you provide `now` to NewFakeClock, it means that every time you use fakeClock.Now(), you’ll get that same initial time. However in this line https://github.com/kubernetes-sigs/kueue/blob/f1efd51f4297cf837bb41f739e6dac546ebcc2ea/pkg/controller/admissionchecks/multikueue/workload_test.go#L811 , we're using time.Now() directly, which means the time could be different each time it’s called. The idea, then, is not only to create a `FakeClock` and propagate it but also to ensure the same time is used in tests to prevent flaky results.

Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/multikueue/workload_test.go:55]: Else, could we replace time.Now() everywhere with just now ??

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi How do i exactly do that ? How will i know which tests are failing ? Please help

You can run tests loccaly manually
- `make verify`
- `make test`
- `make test-integration`

Or you can check logs from test result here https://github.com/kubernetes-sigs/kueue/pull/3498#issuecomment-2466606416. Just press link for each test and you will see result.

File: pkg/controller/admissionchecks/multikueue/workload_test.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/multikueue/workload_test.go:55]: Yeah, that what I mean.

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 3]
```diff
@@ -62,9 +62,6 @@ const (
 var (
 	errInconsistentPodSetAssignments = errors.New("inconsistent podSet assignments")
 )
-var (
-	realClock = clock.RealClock{}
-)
```
Comment (mbobrovskyi) [inline pkg/controller/admissionchecks/provisioning/controller.go:67]: Why did you remove it?

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 4]
```diff
@@ -104,7 +101,6 @@ func NewController(client client.Client, record record.EventRecorder) (*Controll
 		client: client,
 		record: record,
 		helper: helper,
-		clock:  realClock,
```
Comment (mbobrovskyi) [inline pkg/controller/admissionchecks/provisioning/controller.go:107]: And this?

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 3 above)
Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/provisioning/controller.go:67]: This is nowhere being used.

Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/provisioning/controller.go:67]: The intellisence pointed it out

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 4 above)
Comment (TusharMohapatra07) [reply pkg/controller/admissionchecks/provisioning/controller.go:107]: Should i add it back ?

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 3 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go:67]: See comment below.

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 4 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go:107]: I don't know. If it's not need just explain :)

Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go:107]: But I think this might be the reason why tests aren't working.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Iam getting this suggestion from intellisense. 
image
Was it already like that or is it my doing ?

**Comment by mbobrovskyi** [comment]
Yeah, we can optimize it as well, but I'd propose doing it in a separate PR.

File: pkg/workload/admissionchecks.go [hunk 5]
```diff
@@ -87,27 +87,27 @@ func FindAdmissionCheck(checks []kueue.AdmissionCheckState, checkName string) *k
...
-			LastTransitionTime: metav1.NewTime(time.Now()),
+			LastTransitionTime: metav1.NewTime(now),
 			Message:            "Reset to Pending after eviction. Previously: " + string(checks[i].State),
 		}
 	}
 }
 
 // SetAdmissionCheckState - adds or updates newCheck in the provided checks list.
-func SetAdmissionCheckState(checks *[]kueue.AdmissionCheckState, newCheck kueue.AdmissionCheckState) {
+func SetAdmissionCheckState(checks *[]kueue.AdmissionCheckState, newCheck kueue.AdmissionCheckState, now time.Time) {
```
Comment (mbobrovskyi) [inline pkg/workload/admissionchecks.go:103]: Hmm. Maybe these changes aren't necessary We could propagate time in `newCheck kueue.AdmissionCheckState`

https://github.com/kubernetes-sigs/kueue/blob/0c9889edbd3f25b115db4268ac8d646d5a58aad2/pkg/workload/admissionchecks.go#L109-L111

WDYT?

**Comment by mbobrovskyi** [comment]
To fix last test you can execute `make lint-fix`.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi I've fixed the linting. Please give your opinion on above review.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi I've fixed the linting. Please give your opinion on above review.

Thanks! Overall LGTM. Only one case that I mentioned here https://github.com/kubernetes-sigs/kueue/pull/3498#discussion_r1835635000. I think we can revert changes for this function.

File: pkg/workload/admissionchecks.go (hunk 5 above)
Comment (mbobrovskyi) [reply pkg/workload/admissionchecks.go:103]: Please check this case if it's really need.

**Comment by TusharMohapatra07** [comment]
https://github.com/kubernetes-sigs/kueue/pull/3498/files/4c46de9f8b25fedc8180e0ec7cbe2a7296d874e4..a46f9c59bedeb9987c59f9e855449002a7361035#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6 I mean, We could pass in clock.Now() wherever possible and metav1.newTime() anyhow needs it

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi or how about we revert it for now, later we'll have a deep discussion on this.

**Comment by mbobrovskyi** [comment]
> https://github.com/kubernetes-sigs/kueue/pull/3498/files/4c46de9f8b25fedc8180e0ec7cbe2a7296d874e4..a46f9c59bedeb9987c59f9e855449002a7361035#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6 I mean, We could pass in clock.Now() wherever possible (see the above changes, i've passed in w.clock.Now() and so on)and metav1.newTime() anyhow needs it. Reverting it would leave us with two time.Now() instances

Yeah, you right it's using time.Now(), but only if we didn't set `LastTransitionTime`. In which cases does that happen?

**Comment by TusharMohapatra07** [comment]
> > https://github.com/kubernetes-sigs/kueue/pull/3498/files/4c46de9f8b25fedc8180e0ec7cbe2a7296d874e4..a46f9c59bedeb9987c59f9e855449002a7361035#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6 I mean, We could pass in clock.Now() wherever possible (see the above changes, i've passed in w.clock.Now() and so on)and metav1.newTime() anyhow needs it. Reverting it would leave us with two time.Now() instances
> 
> Yeah, you right it's using time.Now(), but only if we didn't set `LastTransitionTime`. In which cases does that happen?

As of now, There's no such case. So I've reverted the changes to `ResetChecksOnEviction`

**Comment by mbobrovskyi** [comment]
> > > https://github.com/kubernetes-sigs/kueue/pull/3498/files/4c46de9f8b25fedc8180e0ec7cbe2a7296d874e4..a46f9c59bedeb9987c59f9e855449002a7361035#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6 I mean, We could pass in clock.Now() wherever possible (see the above changes, i've passed in w.clock.Now() and so on)and metav1.newTime() anyhow needs it. Reverting it would leave us with two time.Now() instances
> > 
> > 
> > Yeah, you right it's using time.Now(), but only if we didn't set `LastTransitionTime`. In which cases does that happen?
> 
> As of now, There's no such case. So, I've reverted the changes to `ResetChecksOnEviction`

I’d like to investigate this case more deeper. From what I see, there are many places in the tests where LastTransitionTime isn’t set, so it should be executing. But maybe instead of propagating time, we just need to update the tests.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Can you please give me an example of where we're setting and where we're not ?? This would really help my thought process as now i'm really confused how it works

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi I think this one need a PR for itself. It'll be easy to cleanup if i mess something . What's your take ?

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi I think this one need a PR for itself. It'll be easy to cleanup if i mess something and also we could keep the discussion specifically for this. What's your take ?

Yes, agree, I think it's better to move it to separate PR and dive deeper to understand how it's working.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Yes, I'll raise a PR regarding this tomorrow. Let's continue this over there :)

**Comment by mbobrovskyi** [comment]
Please also cleanup PR for changes that not need. For example, this one https://github.com/kubernetes-sigs/kueue/pull/3498/files#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6R77-R92.

**Comment by TusharMohapatra07** [comment]
> Please also cleanup PR for changes that not need. For example, this one https://github.com/kubernetes-sigs/kueue/pull/3498/files#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6R77-R92.

Actually, I've used those to implement options in `newWlReconciler`. We're passing in fakeClock using `WithClock` function inside the tests. Is that not needed ?

**Comment by mbobrovskyi** [comment]
But you are not using this clock, right?

**Comment by mbobrovskyi** [comment]
Ah, I see. You missed

https://github.com/kubernetes-sigs/kueue/pull/3498/files#diff-a09b1d2caac9c60d2112a557a72bb2495f52f4c9ac6d161c3b30f1a3a6be24b6R234

**Review by mbobrovskyi — CHANGES_REQUESTED** [review]
Also please check:

https://github.com/kubernetes-sigs/kueue/blob/f0356872b28177893fed8a0db2439c696a699c68/pkg/controller/admissionchecks/multikueue/workload.go#L103

https://github.com/kubernetes-sigs/kueue/blob/f0356872b28177893fed8a0db2439c696a699c68/pkg/controller/admissionchecks/multikueue/workload.go#L413

**Comment by TusharMohapatra07** [comment]
> Also please check:
> 
> https://github.com/kubernetes-sigs/kueue/blob/f0356872b28177893fed8a0db2439c696a699c68/pkg/controller/admissionchecks/multikueue/workload.go#L103

@mbobrovskyi  `wlGroup` doesn't have clock in it. Should i propogate it ?

**Comment by mbobrovskyi** [comment]
> > Also please check:
> > https://github.com/kubernetes-sigs/kueue/blob/f0356872b28177893fed8a0db2439c696a699c68/pkg/controller/admissionchecks/multikueue/workload.go#L103
> 
> @mbobrovskyi `wlGroup` doesn't have clock in it. Should i propogate it ? Also, this is the case that we were discussing yesterday (About creating a seperate PR) !!

No, this is a different case. But you're right - let's fix it in a separate PR. I think we can avoid using time.Now() in this function.

**Comment by TusharMohapatra07** [comment]
> No, this is a different case. But you're right - let's fix it in a separate PR. I think we can avoid using time.Now() in this function.
can we use options ??

**Comment by mbobrovskyi** [comment]
> > No, this is a different case. But you're right - let's fix it in a separate PR. I think we can avoid using time.Now() in this function.
> 
> can we use options ??

Options are useful when we have a constructor, but in this case, we don't. I would propose to avoid using `time.Now()` on this function. We just need to slightly change the logic of this function.

**Comment by TusharMohapatra07** [comment]
> 
> Options are useful when we have a constructor, but in this case, we don't. I would propose to avoid using `time.Now()` on this function. We just need to slightly change the logic of this function.

Okay, Let's do it in a seperate PR

**Comment by TusharMohapatra07** [comment]
TODO: In a seperate PR 
 Issue 1: https://github.com/kubernetes-sigs/kueue/pull/3498#issuecomment-2466811509 (`ResetChecksOnEviction`)
 Issue 2: https://github.com/kubernetes-sigs/kueue/pull/3498#issuecomment-2467479081 (`FirstReserving`)
 
 (This is for my own reference. I'll later add the pr once created beside each so that it's easy to track things)

**Comment by mbobrovskyi** [comment]
Please also fix https://github.com/kubernetes-sigs/kueue/blob/f0356872b28177893fed8a0db2439c696a699c68/pkg/controller/admissionchecks/multikueue/workload.go#L413.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi Done

**Comment by mbobrovskyi** [comment]
> Issue 2: https://github.com/kubernetes-sigs/kueue/pull/3498#issuecomment-2467479081 (FirstReserving)

Looks like it's not very hard case. I think this

```
func (g *wlGroup) FirstReserving() (bool, string) {
	found := false
	bestMatch := ""
	var bestTime time.Time
	for remote, wl := range g.remotes {
		if wl == nil {
			continue
		}
		c := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadQuotaReserved)
		if c != nil && c.Status == metav1.ConditionTrue && (!found || bestTime.IsZero() || c.LastTransitionTime.Time.Before(bestTime)) {
			found = true
			bestMatch = remote
			bestTime = c.LastTransitionTime.Time
		}
	}
	return found, bestMatch
}
```

should work.

**Comment by mbobrovskyi** [comment]
/lgtm 

Amazing! Thank you so much!

**Review by mimowo — COMMENTED** [review]
/approve
Thanks!
